Single-shell Free Water DTI (Object Oriented)

mvgolub opened the pull request:
This is my third pull request for single-shell FW-DTI. The previous PR's (#1603 and #1744) were abandoned because many changes to the code were made since then.

The main difference of this code is that it was implemented with objects and classes, to be consistent with Dipy's implementations of other reconstruction models; also all functions were written inside the existing `dipy/reconst/fwdti.py` module, instead of being implemented in a separate file.

This code is based on a regularized gradient descent approach proposed by [1], but instead uses an euclidean metric for the regularization, as proposed in [2]. Additionally, a more novel initialization method for the FW fraction was implemented, proposed in [3].

The code is not ready to be merged, needs to be reviewed @RafaelNH. When reviewing, please **ignore all commits made before 2020**, I accidentally branched off from an old feature branch and the old outdated commits appear in this PR, I did not know how to fix it (begginer's mistake, sorry).

**TODO:**
1) Add unit tests
2) Add a wrapper to process data by chunks. Currently the code is not optimized to run on large data sets. This code is vectorized, processing the entire data at once. Unlike standard DTI and multi-shell FW-DTI, there is no single-voxel fit and it is not possible to process linearized / flattened chunks of data, because this method requires spatial derivatives to perform regularization, only possible on 3D / 2D multi-voxel data. Maybe it is possible?
3) Add more theory in DOC strings
4) Correct PEP8 errors

**References:**
[1] Pasternak, O., Sochen, N., Gur, Y., Intrator, N., & Assaf, Y. (2009). Free water elimination and mapping from diffusion MRI. Magnetic Resonance in Medicine: An Official Journal of the International Society for Magnetic Resonance in Medicine, 62(3), 717-730. (https://onlinelibrary.wiley.com/doi/full/10.1002/mrm.22055)

[2] Pasternak, O., Maier-Hein, K., Baumgartner, C., Shenton, M. E., Rathi, Y., & Westin, C. F. (2014). The estimation of free-water corrected diffusion tensors. In Visualization and Processing of Tensors and Higher Order Descriptors for Multi-valued Data (pp. 249-270). Springer, Berlin, Heidelberg. (https://link.springer.com/chapter/10.1007/978-3-642-54301-2_11)

[3] Ismail, A. A. O., Parker, D., Hernandez-Fernandez, M., Brem, S., Alexander, S., Pasternak, O., ... & Verma, R. (2018, September). Characterizing peritumoral tissue using DTI-Based free water elimination. In International MICCAI Brainlesion Workshop (pp. 123-131). Springer, Cham. (https://www.hal.inserm.fr/inserm-01867347/file/BrainLes-FWE-2018.pdf)

skoudoro commented:
What a Huge PR for a promising contribution, thank you @mvgolub! and let us know when it is ready for a review

Yshoreway commented:
Thanks for your great contributions. I use a single-shell free water DTI to test your code and save the property fw as a nifti file. I open the fw nifti file in Mricron, it looks like 
![批注 2020-05-18 104657](https://user-images.githubusercontent.com/22960923/82226797-f189bf00-98f4-11ea-9418-238a3abe5861.jpg).  Could you give me some advice to fix the problem?

mvgolub commented:
> Thanks for your great contributions. I use a single-shell free water DTI to test your code and save the property fw as a nifti file. I open the fw nifti file in Mricron, it looks like
> ![批注 2020-05-18 104657](https://user-images.githubusercontent.com/22960923/82226797-f189bf00-98f4-11ea-9418-238a3abe5861.jpg). Could you give me some advice to fix the problem?

Hi @Yshoreway, I think the problem is that you need to rescale the b-values to the units of ms/micrometer^2 (i.e. a bval of 1000 s/mm^2  should become 1) before you construct the `GradientTable` class instance and pass it to the `BeltramiModel`class (I should have mentioned this in the PR request, sorry).

This is because the single-shell fitting method uses finite differences that can become very small depending on the chosen learning rate and lead to numerical instabilities, to solve this, I set the internal value of the free water diffusion constant to `Dw = 3` (instead of `3*10**-3`), and the the input b-values must be rescaled.

Hope this solves the problem, if not, feel free to contact me by email.

Yshoreway commented:
<!--
/* Font Definitions */
@font-face
	{font-family:"Cambria Math";
	panose-1:2 4 5 3 5 4 6 3 2 4;}
@font-face
	{font-family:DengXian;
	panose-1:2 1 6 0 3 1 1 1 1 1;}
@font-face
	{font-family:Calibri;
	panose-1:2 15 5 2 2 2 4 3 2 4;}
@font-face
	{font-family:DengXian;
	panose-1:2 1 6 0 3 1 1 1 1 1;}
/* Style Definitions */
p.MsoNormal, li.MsoNormal, div.MsoNormal
	{margin:0cm;
	margin-bottom:.0001pt;
	text-align:justify;
	text-justify:inter-ideograph;
	font-size:10.5pt;
	font-family:"Calibri",sans-serif;}
a:link, span.MsoHyperlink
	{mso-style-priority:99;
	color:blue;
	text-decoration:underline;}
code
	{mso-style-priority:99;
	font-family:"Courier New";}
.MsoChpDefault
	{mso-style-type:export-only;}
/* Page Definitions */
@page WordSection1
	{size:612.0pt 792.0pt;
	margin:72.0pt 90.0pt 72.0pt 90.0pt;}
div.WordSection1
	{page:WordSection1;}
-->Thank you for your reply. I will try and see. Thanks!!!  发件人: Marc发送时间: Monday, May 18, 2020 11:31 AM收件人: dipy/dipy抄送: Shoreway; Mention主题: Re: [dipy/dipy] Single-shell Free Water DTI (Object Oriented) (#2087) Thanks for your great contributions. I use a single-shell free water DTI to test your code and save the property fw as a nifti file. I open the fw nifti file in Mricron, it looks like. Could you give me some advice to fix the problem?Hi @Yshoreway, I think the problem is that you need to rescale the b-values to the units of ms/micrometer^2 (i.e. a bval of 1000 s/mm^2 should become 1) before you construct the GradientTable class instance and pass it to the BeltramiModelclass (I should have mentioned this in the PR request, sorry).This is because the single-shell fitting method uses finite differences that can become very small depending on the chosen learning rate and lead to numerical instabilities, to solve this, I set the internal value of the free water diffusion constant to Dw = 3 (instead of 3*10**-3), and the the input b-values must be rescaled.Hope this solves the problem, if not, feel free to contact me by email.—You are receiving this because you were mentioned.Reply to this email directly, view it on GitHub, or unsubscribe. 

Yshoreway commented:
<!--
/* Font Definitions */
@font-face
	{font-family:"Cambria Math";
	panose-1:2 4 5 3 5 4 6 3 2 4;}
@font-face
	{font-family:DengXian;
	panose-1:2 1 6 0 3 1 1 1 1 1;}
@font-face
	{font-family:Calibri;
	panose-1:2 15 5 2 2 2 4 3 2 4;}
@font-face
	{font-family:DengXian;
	panose-1:2 1 6 0 3 1 1 1 1 1;}
/* Style Definitions */
p.MsoNormal, li.MsoNormal, div.MsoNormal
	{margin:0cm;
	margin-bottom:.0001pt;
	text-align:justify;
	text-justify:inter-ideograph;
	font-size:10.5pt;
	font-family:"Calibri",sans-serif;}
a:link, span.MsoHyperlink
	{mso-style-priority:99;
	color:blue;
	text-decoration:underline;}
code
	{mso-style-priority:99;
	font-family:"Courier New";}
.MsoChpDefault
	{mso-style-type:export-only;}
/* Page Definitions */
@page WordSection1
	{size:612.0pt 792.0pt;
	margin:72.0pt 90.0pt 72.0pt 90.0pt;}
div.WordSection1
	{page:WordSection1;}
-->Hi, After rescaling the b-values to the units of ms/micrometer^2, it seems I got a satisfying free water map. It looks like this, which is attached. But I got a warning, “C:\ProgramData\Anaconda3\lib\site-packages\dipy\reconst\fwdti.py:1150: RuntimeWarning: invalid value encountered in true_divide axis=-1) / k”. It happens in  line 1158,  self.flat_cost[..., 0] = np.sum((Amodel - self.flat_attenuations)**2, axis=-1) /k.I print the value, obviously this warning is caused by 0/0.  “np.sum((Amodel - self.flat_attenuations)**2,axis=-1) is [0. 0. 0. ... 0. 0. 0.],k is 0” . Does this affect the free water map result? How to fix this bug? Thanks. 发件人: Marc发送时间: Monday, May 18, 2020 11:31 AM收件人: dipy/dipy抄送: Shoreway; Mention主题: Re: [dipy/dipy] Single-shell Free Water DTI (Object Oriented) (#2087) Thanks for your great contributions. I use a single-shell free water DTI to test your code and save the property fw as a nifti file. I open the fw nifti file in Mricron, it looks like. Could you give me some advice to fix the problem?Hi @Yshoreway, I think the problem is that you need to rescale the b-values to the units of ms/micrometer^2 (i.e. a bval of 1000 s/mm^2 should become 1) before you construct the GradientTable class instance and pass it to the BeltramiModelclass (I should have mentioned this in the PR request, sorry).This is because the single-shell fitting method uses finite differences that can become very small depending on the chosen learning rate and lead to numerical instabilities, to solve this, I set the internal value of the free water diffusion constant to Dw = 3 (instead of 3*10**-3), and the the input b-values must be rescaled.Hope this solves the problem, if not, feel free to contact me by email.—You are receiving this because you were mentioned.Reply to this email directly, view it on GitHub, or unsubscribe. 

Yshoreway commented:
Hi, after rescaling the b-values to the units of ms/micrometer^2, it seems I got a satisfying free water map. It looks like this. But I got a warning, “C:\ProgramData\Anaconda3\lib\site-packages\dipy\reconst\fwdti.py:1150: RuntimeWarning: invalid value encountered in true_divide axis=-1) / k”. It happens in  line 1158,  self.flat_cost[..., 0] = np.sum((Amodel - self.flat_attenuations)**2, axis=-1) /k.
I print the values, obviously this warning is caused by 0/0.  “np.sum((Amodel - self.flat_attenuations)**2,axis=-1) is [0. 0. 0. ... 0. 0. 0.],k is 0” . Does this affect the free water map result? How to fix this bug? Thanks.
![批注 2020-05-18 114619](https://user-images.githubusercontent.com/22960923/82234101-eb004500-98fe-11ea-98b1-7a17fca4f1dc.jpg)

mvgolub commented:
> Hi, after rescaling the b-values to the units of ms/micrometer^2, it seems I got a satisfying free water map. It looks like this. But I got a warning, “C:\ProgramData\Anaconda3\lib\site-packages\dipy\reconst\fwdti.py:1150: RuntimeWarning: invalid value encountered in true_divide axis=-1) / k”. It happens in line 1158, self.flat_cost[..., 0] = np.sum((Amodel - self.flat_attenuations)**2, axis=-1) /k.
> I print the values, obviously this warning is caused by 0/0. “np.sum((Amodel - self.flat_attenuations)**2,axis=-1) is [0. 0. 0. ... 0. 0. 0.],k is 0” . Does this affect the free water map result? How to fix this bug? Thanks.
> ![批注 2020-05-18 114619](https://user-images.githubusercontent.com/22960923/82234101-eb004500-98fe-11ea-98b1-7a17fca4f1dc.jpg)

It appears to be a problem in the cost function, technically the output of this function is not used to update the parameters, I only implemented it for debugging purposes. In order to fix this, I have to take a closer look later. But if you want to run the code without the warning and make sure the parameters are not affected, try commenting the line 1219, where the cost function is called.

Yshoreway commented:
Thanks. Btw, do I need to rescale the b-values to the units of ms/micrometer^2 when I use multi-shell free water mapping?

mvgolub commented:
> Thanks. Btw, do I need to rescale the b-values to the units of ms/micrometer^2 when I use multi-shell free water mapping?

Only if you use the `BeltramiModel` class to fit the data. However, if you have multi-shell data, you should use the `FreeWaterTensorModel` class because it was already tested and fully integrated into DIPY. Best regards.

davidjouellette commented:
Fantastic project, @mvgolub ! I'm new to DIPY and I had a question about the usage: assuming I'm okay with the default initialization, is it as simple as something from the DTI tutorial, e.g.:

import dipy.reconst.fwdti as fwdti (etc, etc)

tmodel = fwdti.BeltramiModel(gtab, 'MD')
tfit = tmodel.fit(data)

I saw that there were additional inputs for the model initialization and gradient_descent but I wasn't sure if they were required. For example you can specify the 'MD' method in the model but fraction_init_MD it isn't called anywhere in fwdti, so I assume I have to call it myself to estimate the fractions and feed into the model? The above code gave me an "eigenvalues did not converge" error from numpy so I must be missing something.

mvgolub commented:
> Fantastic project, @mvgolub ! I'm new to DIPY and I had a question about the usage: assuming I'm okay with the default initialization, is it as simple as something from the DTI tutorial, e.g.:
> 
> import dipy.reconst.fwdti as fwdti (etc, etc)
> 
> tmodel = fwdti.BeltramiModel(gtab, 'MD')
> tfit = tmodel.fit(data)
> 
> I saw that there were additional inputs for the model initialization and gradient_descent but I wasn't sure if they were required. For example you can specify the 'MD' method in the model but fraction_init_MD it isn't called anywhere in fwdti, so I assume I have to call it myself to estimate the fractions and feed into the model? The above code gave me an "eigenvalues did not converge" error from numpy so I must be missing something.

Thank you. Yes, to use the code it should be as simple as to use other methods from DIPY. One small correction to your example:
`tmodel = fwdti.Model(gtab, init_method='MD')`
`tfit = tmodel.fit(data)`

The string 'MD' is passed as a keyword argument for `init_method`, if you run your example you will get an error. The `init_mehtod` is called at line 1352 of `fwdti.py` depending on the input string: 'MD', 'S0' or 'hybrid'. However, "MD" is already the default method and the `fraction_init_md` function will be called, so you can just run `tmodel = fwdti.BeltramiModel(gtab)`, i.e. no need to estimate your own fractions and feed them to the model.

Some keyword arguments that you could consider are `learning_rate` and `iterations`, for example:
`tmodel = fwdti.BeltramiModel(gtab, learning_rate=0.0005, iterations=200)`
To see other keyword arguments and their defaults check the `gradient_descent` function at line 1763.

Also, consider running `tfit = tmodel.fit(data, mask=your_mask)`, with a previously computed binary mask by you, because this method uses spatial derivatives that can become unstable at the brain borders and propagate, a good mask can help mitigate this problem.

Finally, to extract the your scalar maps you could do:
`fa = tfit.fa`
`fw = tfit.fw`
and so on, just like standard DTI.

davidjouellette commented:
> Thank you @davidjouellette ! Yes, to use the code it should be as simple as to use other methods from DIPY. One small correction to your example:
> `tmodel = fwdti.Model(gtab, init_method='MD')`
> `tfit = tmodel.fit(data)`
> 
> The string 'MD' is passed as a keyword argument for `init_method`, if you run your example you will get an error. The `init_mehtod` is called at line 1352 of `fwdti.py` depending on the input string: 'MD', 'S0' or 'hybrid'. However, "MD" is already the default method and the `fraction_init_md` function will be called, so you can just run `tmodel = fwdti.BeltramiModel(gtab)`, i.e. no need to estimate your own fractions and feed them to the model.
> 
> Some keyword arguments that you could consider are `learning_rate` and `iterations`, for example:
> `tmodel = fwdti.BeltramiModel(gtab, learning_rate=0.0005, iterations=200)`
> To see other keyword arguments and their defaults check the `gradient_descent` function at line 1763.
> 
> Also, consider running `tfit = tmodel.fit(data, mask=your_mask)`, with a previously computed binary mask by you, because this method uses spatial derivatives that can become unstable at the brain borders and propagate, a good mask can help mitigate this problem.
> 
> Finally, to extract the scalar maps you could do:
> `fa = tfit.fa`
> `fw = tfit.fw`
> and so on, just like standard DTI.

Thanks for the quick reply, it makes a lot more sense to me now. However, I'm having an issue with some of the parameter maps. The free water map comes out as expected, but the other common ones (AD, RD, MD) are all set to 1e-6 everywhere in the ROI. If I plot tfit.model_params, it's either all ones or all zeros in the ROI except for the last dimension (the free water). Any idea what could be causing this? If you want to take it off line please email me at (my username).gmail.com, as I expect this is more about my implementation than an issue with the branch.

mvgolub commented:
> 
> 
> > Thank you @davidjouellette ! Yes, to use the code it should be as simple as to use other methods from DIPY. One small correction to your example:
> > `tmodel = fwdti.Model(gtab, init_method='MD')`
> > `tfit = tmodel.fit(data)`
> > The string 'MD' is passed as a keyword argument for `init_method`, if you run your example you will get an error. The `init_mehtod` is called at line 1352 of `fwdti.py` depending on the input string: 'MD', 'S0' or 'hybrid'. However, "MD" is already the default method and the `fraction_init_md` function will be called, so you can just run `tmodel = fwdti.BeltramiModel(gtab)`, i.e. no need to estimate your own fractions and feed them to the model.
> > Some keyword arguments that you could consider are `learning_rate` and `iterations`, for example:
> > `tmodel = fwdti.BeltramiModel(gtab, learning_rate=0.0005, iterations=200)`
> > To see other keyword arguments and their defaults check the `gradient_descent` function at line 1763.
> > Also, consider running `tfit = tmodel.fit(data, mask=your_mask)`, with a previously computed binary mask by you, because this method uses spatial derivatives that can become unstable at the brain borders and propagate, a good mask can help mitigate this problem.
> > Finally, to extract the scalar maps you could do:
> > `fa = tfit.fa`
> > `fw = tfit.fw`
> > and so on, just like standard DTI.
> 
> Thanks for the quick reply, it makes a lot more sense to me now. However, I'm having an issue with some of the parameter maps. The free water map comes out as expected, but the other common ones (AD, RD, MD) are all set to 1e-6 everywhere in the ROI. If I plot tfit.model_params, it's either all ones or all zeros in the ROI except for the last dimension (the free water). Any idea what could be causing this? If you want to take it off line please email me at (my username).gmail.com, as I expect this is more about my implementation than an issue with the branch.

What learning rate did you use? Did you use the default, which is 0.01? Because now that I think of it, the default learning rate might be too high for most cases, try using "learning_rate=0.0001"

Did you rescale the b-values before constructing the gtab (i.e a bval of 1000 becomes 1), like I suggested in another reply?

Can you try running the code with one of DIPY's datasets used for the tutorials? For example, try to process some slices of the the "stanford_hardi" dataset  (the one used in the standard DTI tutorial) and check if you have the same problem. My gmail is marc.golub93 for further questions if you like.

mvgolub commented:
Hi @mbergamino, the issue is that when you construct the gradients table, Dipy's function`GradientTable` interprets the rescaled bvals as very small and cuts them down to 0. Although you don't show it in your code example, I'm assuming that you are using `gtab = GradientTable(bvals_rescaled, bvecs)` to make your gtab, then I think that this should fix the problem:

`bvals_rescaled = bvals * 10**-3`
`gtab = GradientTable(bvals_rescaled, bvecs, b0_threshold=0)`
`tmodel = ...`

By specifying that `b0_threshold=0` you are telling `GradientTable` that bvals in the scale of unity should not be cropped to 0.

Regards,

Marc

mbergamino commented:
Hi @mvgolub
Thank you for your reply.
I have changed my script and now I have the free-water FA.
Only a question:
The free-water map looks fine. 
I have some doubts about the free-water FA.

I have also created the standard FA. In some brain locations  I have (e.g. ventricles):
FW = 1.0
Standard FA = 0.05552
FW-FA = 0.9999

In theory, I should have the FW-FA very close to 0 in those locations.
Did you see this issue before?

Thanks,
Maurizio

umgpy commented:
Hi, has this been merged with the main branch? I would love to use it for my analysis.
Much thanks!

skoudoro commented:
Hi @mvgolub and @RafaelNH,

Can we get a quick feedback about this work?

What is the ETA for completing your to-do listed above? How can we help?

mvgolub commented:
> Hi @mvgolub and @RafaelNH,
> 
> Can we get a quick feedback about this work?
> 
> What is the ETA for completing your to-do listed above? How can we help?

pep8speaks commented:
Hello @mvgolub, Thank you for updating !

* In the file [`dipy/reconst/fwdti.py`](https://github.com/dipy/dipy/blob/42fd5227b512b640bd917e58b4fbc9fb0e1333c1/dipy/reconst/fwdti.py):

> [Line 829:54](https://github.com/dipy/dipy/blob/42fd5227b512b640bd917e58b4fbc9fb0e1333c1/dipy/reconst/fwdti.py#L829): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 859:1](https://github.com/dipy/dipy/blob/42fd5227b512b640bd917e58b4fbc9fb0e1333c1/dipy/reconst/fwdti.py#L859): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 900:1](https://github.com/dipy/dipy/blob/42fd5227b512b640bd917e58b4fbc9fb0e1333c1/dipy/reconst/fwdti.py#L900): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 947:5](https://github.com/dipy/dipy/blob/42fd5227b512b640bd917e58b4fbc9fb0e1333c1/dipy/reconst/fwdti.py#L947): [E303](https://duckduckgo.com/?q=pep8%20E303) too many blank lines (2)
> [Line 984:5](https://github.com/dipy/dipy/blob/42fd5227b512b640bd917e58b4fbc9fb0e1333c1/dipy/reconst/fwdti.py#L984): [E303](https://duckduckgo.com/?q=pep8%20E303) too many blank lines (2)
> [Line 1021:5](https://github.com/dipy/dipy/blob/42fd5227b512b640bd917e58b4fbc9fb0e1333c1/dipy/reconst/fwdti.py#L1021): [E303](https://duckduckgo.com/?q=pep8%20E303) too many blank lines (2)
> [Line 1028:5](https://github.com/dipy/dipy/blob/42fd5227b512b640bd917e58b4fbc9fb0e1333c1/dipy/reconst/fwdti.py#L1028): [E303](https://duckduckgo.com/?q=pep8%20E303) too many blank lines (2)
> [Line 1055:52](https://github.com/dipy/dipy/blob/42fd5227b512b640bd917e58b4fbc9fb0e1333c1/dipy/reconst/fwdti.py#L1055): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 1097:1](https://github.com/dipy/dipy/blob/42fd5227b512b640bd917e58b4fbc9fb0e1333c1/dipy/reconst/fwdti.py#L1097): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 1104:26](https://github.com/dipy/dipy/blob/42fd5227b512b640bd917e58b4fbc9fb0e1333c1/dipy/reconst/fwdti.py#L1104): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 1105:1](https://github.com/dipy/dipy/blob/42fd5227b512b640bd917e58b4fbc9fb0e1333c1/dipy/reconst/fwdti.py#L1105): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 1115:5](https://github.com/dipy/dipy/blob/42fd5227b512b640bd917e58b4fbc9fb0e1333c1/dipy/reconst/fwdti.py#L1115): [E303](https://duckduckgo.com/?q=pep8%20E303) too many blank lines (2)
> [Line 1128:32](https://github.com/dipy/dipy/blob/42fd5227b512b640bd917e58b4fbc9fb0e1333c1/dipy/reconst/fwdti.py#L1128): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 1131:36](https://github.com/dipy/dipy/blob/42fd5227b512b640bd917e58b4fbc9fb0e1333c1/dipy/reconst/fwdti.py#L1131): [E127](https://duckduckgo.com/?q=pep8%20E127) continuation line over-indented for visual indent
> [Line 1146:5](https://github.com/dipy/dipy/blob/42fd5227b512b640bd917e58b4fbc9fb0e1333c1/dipy/reconst/fwdti.py#L1146): [E303](https://duckduckgo.com/?q=pep8%20E303) too many blank lines (2)
> [Line 1151:36](https://github.com/dipy/dipy/blob/42fd5227b512b640bd917e58b4fbc9fb0e1333c1/dipy/reconst/fwdti.py#L1151): [E127](https://duckduckgo.com/?q=pep8%20E127) continuation line over-indented for visual indent
> [Line 1163:5](https://github.com/dipy/dipy/blob/42fd5227b512b640bd917e58b4fbc9fb0e1333c1/dipy/reconst/fwdti.py#L1163): [E303](https://duckduckgo.com/?q=pep8%20E303) too many blank lines (2)
> [Line 1197:67](https://github.com/dipy/dipy/blob/42fd5227b512b640bd917e58b4fbc9fb0e1333c1/dipy/reconst/fwdti.py#L1197): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 1222:5](https://github.com/dipy/dipy/blob/42fd5227b512b640bd917e58b4fbc9fb0e1333c1/dipy/reconst/fwdti.py#L1222): [E303](https://duckduckgo.com/?q=pep8%20E303) too many blank lines (2)
> [Line 1282:71](https://github.com/dipy/dipy/blob/42fd5227b512b640bd917e58b4fbc9fb0e1333c1/dipy/reconst/fwdti.py#L1282): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 1315:30](https://github.com/dipy/dipy/blob/42fd5227b512b640bd917e58b4fbc9fb0e1333c1/dipy/reconst/fwdti.py#L1315): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ':'
> [Line 1318:29](https://github.com/dipy/dipy/blob/42fd5227b512b640bd917e58b4fbc9fb0e1333c1/dipy/reconst/fwdti.py#L1318): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ':'
> [Line 1321:5](https://github.com/dipy/dipy/blob/42fd5227b512b640bd917e58b4fbc9fb0e1333c1/dipy/reconst/fwdti.py#L1321): [E303](https://duckduckgo.com/?q=pep8%20E303) too many blank lines (2)
> [Line 1326:5](https://github.com/dipy/dipy/blob/42fd5227b512b640bd917e58b4fbc9fb0e1333c1/dipy/reconst/fwdti.py#L1326): [E303](https://duckduckgo.com/?q=pep8%20E303) too many blank lines (2)
> [Line 1355:40](https://github.com/dipy/dipy/blob/42fd5227b512b640bd917e58b4fbc9fb0e1333c1/dipy/reconst/fwdti.py#L1355): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 1363:46](https://github.com/dipy/dipy/blob/42fd5227b512b640bd917e58b4fbc9fb0e1333c1/dipy/reconst/fwdti.py#L1363): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 1364:46](https://github.com/dipy/dipy/blob/42fd5227b512b640bd917e58b4fbc9fb0e1333c1/dipy/reconst/fwdti.py#L1364): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 1365:46](https://github.com/dipy/dipy/blob/42fd5227b512b640bd917e58b4fbc9fb0e1333c1/dipy/reconst/fwdti.py#L1365): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 1381:1](https://github.com/dipy/dipy/blob/42fd5227b512b640bd917e58b4fbc9fb0e1333c1/dipy/reconst/fwdti.py#L1381): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 1383:1](https://github.com/dipy/dipy/blob/42fd5227b512b640bd917e58b4fbc9fb0e1333c1/dipy/reconst/fwdti.py#L1383): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 1421:5](https://github.com/dipy/dipy/blob/42fd5227b512b640bd917e58b4fbc9fb0e1333c1/dipy/reconst/fwdti.py#L1421): [E303](https://duckduckgo.com/?q=pep8%20E303) too many blank lines (2)
> [Line 1426:5](https://github.com/dipy/dipy/blob/42fd5227b512b640bd917e58b4fbc9fb0e1333c1/dipy/reconst/fwdti.py#L1426): [E303](https://duckduckgo.com/?q=pep8%20E303) too many blank lines (2)
> [Line 1431:5](https://github.com/dipy/dipy/blob/42fd5227b512b640bd917e58b4fbc9fb0e1333c1/dipy/reconst/fwdti.py#L1431): [E303](https://duckduckgo.com/?q=pep8%20E303) too many blank lines (2)
> [Line 1436:5](https://github.com/dipy/dipy/blob/42fd5227b512b640bd917e58b4fbc9fb0e1333c1/dipy/reconst/fwdti.py#L1436): [E303](https://duckduckgo.com/?q=pep8%20E303) too many blank lines (2)
> [Line 1441:5](https://github.com/dipy/dipy/blob/42fd5227b512b640bd917e58b4fbc9fb0e1333c1/dipy/reconst/fwdti.py#L1441): [E303](https://duckduckgo.com/?q=pep8%20E303) too many blank lines (2)
> [Line 1446:5](https://github.com/dipy/dipy/blob/42fd5227b512b640bd917e58b4fbc9fb0e1333c1/dipy/reconst/fwdti.py#L1446): [E303](https://duckduckgo.com/?q=pep8%20E303) too many blank lines (2)
> [Line 1451:5](https://github.com/dipy/dipy/blob/42fd5227b512b640bd917e58b4fbc9fb0e1333c1/dipy/reconst/fwdti.py#L1451): [E303](https://duckduckgo.com/?q=pep8%20E303) too many blank lines (2)
> [Line 1490:31](https://github.com/dipy/dipy/blob/42fd5227b512b640bd917e58b4fbc9fb0e1333c1/dipy/reconst/fwdti.py#L1490): [E203](https://duckduckgo.com/?q=pep8%20E203) whitespace before ','
> [Line 1492:45](https://github.com/dipy/dipy/blob/42fd5227b512b640bd917e58b4fbc9fb0e1333c1/dipy/reconst/fwdti.py#L1492): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 1541:67](https://github.com/dipy/dipy/blob/42fd5227b512b640bd917e58b4fbc9fb0e1333c1/dipy/reconst/fwdti.py#L1541): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 1548:40](https://github.com/dipy/dipy/blob/42fd5227b512b640bd917e58b4fbc9fb0e1333c1/dipy/reconst/fwdti.py#L1548): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 1550:81](https://github.com/dipy/dipy/blob/42fd5227b512b640bd917e58b4fbc9fb0e1333c1/dipy/reconst/fwdti.py#L1550): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (89 > 80 characters)
> [Line 1620:39](https://github.com/dipy/dipy/blob/42fd5227b512b640bd917e58b4fbc9fb0e1333c1/dipy/reconst/fwdti.py#L1620): [E225](https://duckduckgo.com/?q=pep8%20E225) missing whitespace around operator
> [Line 1620:49](https://github.com/dipy/dipy/blob/42fd5227b512b640bd917e58b4fbc9fb0e1333c1/dipy/reconst/fwdti.py#L1620): [E225](https://duckduckgo.com/?q=pep8%20E225) missing whitespace around operator
> [Line 1621:39](https://github.com/dipy/dipy/blob/42fd5227b512b640bd917e58b4fbc9fb0e1333c1/dipy/reconst/fwdti.py#L1621): [E225](https://duckduckgo.com/?q=pep8%20E225) missing whitespace around operator
> [Line 1621:49](https://github.com/dipy/dipy/blob/42fd5227b512b640bd917e58b4fbc9fb0e1333c1/dipy/reconst/fwdti.py#L1621): [E225](https://duckduckgo.com/?q=pep8%20E225) missing whitespace around operator
> [Line 1623:46](https://github.com/dipy/dipy/blob/42fd5227b512b640bd917e58b4fbc9fb0e1333c1/dipy/reconst/fwdti.py#L1623): [E225](https://duckduckgo.com/?q=pep8%20E225) missing whitespace around operator
> [Line 1623:56](https://github.com/dipy/dipy/blob/42fd5227b512b640bd917e58b4fbc9fb0e1333c1/dipy/reconst/fwdti.py#L1623): [E225](https://duckduckgo.com/?q=pep8%20E225) missing whitespace around operator
> [Line 1742:1](https://github.com/dipy/dipy/blob/42fd5227b512b640bd917e58b4fbc9fb0e1333c1/dipy/reconst/fwdti.py#L1742): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 1818:67](https://github.com/dipy/dipy/blob/42fd5227b512b640bd917e58b4fbc9fb0e1333c1/dipy/reconst/fwdti.py#L1818): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 1857:21](https://github.com/dipy/dipy/blob/42fd5227b512b640bd917e58b4fbc9fb0e1333c1/dipy/reconst/fwdti.py#L1857): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ':'
> [Line 1858:21](https://github.com/dipy/dipy/blob/42fd5227b512b640bd917e58b4fbc9fb0e1333c1/dipy/reconst/fwdti.py#L1858): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ':'
> [Line 1859:21](https://github.com/dipy/dipy/blob/42fd5227b512b640bd917e58b4fbc9fb0e1333c1/dipy/reconst/fwdti.py#L1859): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ':'
> [Line 1860:21](https://github.com/dipy/dipy/blob/42fd5227b512b640bd917e58b4fbc9fb0e1333c1/dipy/reconst/fwdti.py#L1860): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ':'
> [Line 1861:21](https://github.com/dipy/dipy/blob/42fd5227b512b640bd917e58b4fbc9fb0e1333c1/dipy/reconst/fwdti.py#L1861): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ':'
> [Line 1862:35](https://github.com/dipy/dipy/blob/42fd5227b512b640bd917e58b4fbc9fb0e1333c1/dipy/reconst/fwdti.py#L1862): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ':'
> [Line 1863:25](https://github.com/dipy/dipy/blob/42fd5227b512b640bd917e58b4fbc9fb0e1333c1/dipy/reconst/fwdti.py#L1863): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ':'
> [Line 1864:25](https://github.com/dipy/dipy/blob/42fd5227b512b640bd917e58b4fbc9fb0e1333c1/dipy/reconst/fwdti.py#L1864): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ':'
> [Line 1865:29](https://github.com/dipy/dipy/blob/42fd5227b512b640bd917e58b4fbc9fb0e1333c1/dipy/reconst/fwdti.py#L1865): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ':'

mvgolub commented:
> Hi @mvgolub and @RafaelNH,
> 
> Can we get a quick feedback about this work?
> 
> What is the ETA for completing your to-do listed above? How can we help?

Hi @skoudoro , sorry for the long silence on my part. Last time we talked with Rafael, we agreed to remove the "regularization" from this implementation because we concluded it offers little benefit after the initialization for single-shell data, it was also becoming hard to make it compatible with the current multi-shell routine (point 2 of the TODO list).

Since then, I implemented the initialization routine (without reg) in a single function `fernet_iter` that takes single-voxel signals and is decorated with the `multi_voxel_fit`,  and is therefore compatible with the models and classes of the current `reconst.fwdti.py ` module; also modified the `FreeWaterTensorModel` class so that if it receives single-shell data, instead of aborting, it jumps to he single-shell routine. I did this in a separate branch https://github.com/mvgolub/dipy/blob/singleshell-fwdti-noreg/dipy/reconst/fwdti.py.

The docstrings are completed I think, so the only thing missing from the TODO list are the unit tests, which I am bad at designing them, so I will ask Rafael for some help. If you are familiar with the code in `reconst.fwdti.py ` and have any suggestions on how we should approach the unit testing for the function that I added, feel free to share.

skoudoro commented:
Thank you for the feedback and the update @mvgolub.

I will look into it and try to help also. 

Currently, you have a small typo (`+=` instead of `=+`), that's why the tests are failing:

```python
if not enough_b and self.fit_method in (wls_iter, nls_iter):
            mes = "fwDTI requires at least 3 b-values (which can include b=0)"
>           mes =+ " when using the NLS or WLS fit methods"
E           TypeError: bad operand type for unary +: 'str'

```

arokem commented on dipy/reconst/fwdti.py:
```diff
@@ -126,11 +126,28 @@ def __init__(self, gtab, fit_method="NLS", *args, **kwargs):
-        if not enough_b:
+        if not enough_b and self.fit_method in (wls_iter, nls_iter):
             mes = "fwDTI requires at least 3 b-values (which can include b=0)"
+            mes =+ " when using the NLS or WLS fit methods"
```
```suggestion
            mes += " when using the NLS or WLS fit methods"
```

arokem commented:
Fixed that typo in https://github.com/dipy/dipy/pull/2087/commits/7748b366be25bcd9180fa6d6572c7c10eef30d28

txusser commented:
Hi. What is the status of this request? Has it merged? The pull request seems to be open.

Thanks in advance,

skoudoro commented:
Hi @txusser,

it is not ready to be merged. See below what is missing:

- Add unit tests
- Correct PEP8 errors
- multiple reviews round. 

your review and feedback (theory or code) is welcomed! Please let us know if you have any question concerning theory or code.

txusser commented:
I will try to look into it. This is so close to merge and it is a very relevant addition imho, it would be a pity for it to keep hanging.

alrodrig11 commented:
Has this ever been implemented in DIPY? I have single shell data that I would like to analyze using the same method/software as my multishell data.

mbergamino commented:
The algorithm in dipy works only in multi-shell data

On Wed, Oct 11, 2023, 3:46 PM Amanda Rodrigue ***@***.***>
wrote:

> Has this ever been implemented in DIPY? I have single shell data that I
> would like to analyze using the same method/software as my multishell data.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/dipy/dipy/pull/2087#issuecomment-1758512028>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AGEBKOFWJF3F5UIZVL4QOMTX64AQRANCNFSM4LHCYCJQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

alrodrig11 commented:
Yes, but there seems to be a pretty long history of developing the code for single shell data, but this seems to have stalled? Has this project been abandoned?
Thanks,
Amanda
________________________________
From: mbergamino ***@***.***>
Sent: Wednesday, October 11, 2023 4:50 PM
To: dipy/dipy ***@***.***>
Cc: Rodrigue, Amanda ***@***.***>; Comment ***@***.***>
Subject: Re: [dipy/dipy] Single-shell Free Water DTI (Object Oriented) (#2087) [EXTERNAL]

* External Email - Caution *


The algorithm in dipy works only in multi-shell data

On Wed, Oct 11, 2023, 3:46 PM Amanda Rodrigue ***@***.***>
wrote:

> Has this ever been implemented in DIPY? I have single shell data that I
> would like to analyze using the same method/software as my multishell data.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/dipy/dipy/pull/2087#issuecomment-1758512028><https://urldefense.com/v3/__https://github.com/dipy/dipy/pull/2087*issuecomment-1758512028*3E__;IyU!!NZvER7FxgEiBAiR_!t10I1q_z9AIF3baA8tknJNOIw5qKVbRVJ39mPpTJPwBU3z9XWLL_21WC6NKhgZTJfSR8UHu3cARZabYhbWdsq187z_IVTJjuFUu8h5w$>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AGEBKOFWJF3F5UIZVL4QOMTX64AQRANCNFSM4LHCYCJQ><https://urldefense.com/v3/__https://github.com/notifications/unsubscribe-auth/AGEBKOFWJF3F5UIZVL4QOMTX64AQRANCNFSM4LHCYCJQ*3E__;JQ!!NZvER7FxgEiBAiR_!t10I1q_z9AIF3baA8tknJNOIw5qKVbRVJ39mPpTJPwBU3z9XWLL_21WC6NKhgZTJfSR8UHu3cARZabYhbWdsq187z_IVTJju33GkZ4g$>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

—
Reply to this email directly, view it on GitHub<https://urldefense.com/v3/__https://github.com/dipy/dipy/pull/2087*issuecomment-1758517348__;Iw!!NZvER7FxgEiBAiR_!t10I1q_z9AIF3baA8tknJNOIw5qKVbRVJ39mPpTJPwBU3z9XWLL_21WC6NKhgZTJfSR8UHu3cARZabYhbWdsq187z_IVTJjuDGfOmTQ$>, or unsubscribe<https://urldefense.com/v3/__https://github.com/notifications/unsubscribe-auth/AN347Y6LZJM2UN5BGYYKKHTX64A75ANCNFSM4LHCYCJQ__;!!NZvER7FxgEiBAiR_!t10I1q_z9AIF3baA8tknJNOIw5qKVbRVJ39mPpTJPwBU3z9XWLL_21WC6NKhgZTJfSR8UHu3cARZabYhbWdsq187z_IVTJjuCjCMXag$>.
You are receiving this because you commented.Message ID: ***@***.***>

skoudoro commented:
Hi @alrodrig11,

As explained [above](https://github.com/dipy/dipy/pull/2087#issuecomment-1694692568), this PR needs a bit of work.

it is not ready to be merged. See below what is missing:

- Add unit tests
- Correct PEP8 errors
- multiple reviews round.

your review and feedback (theory or code) is welcomed! Please let us know if you have any question concerning theory or code. You can easily get this branch, run the model and give us your opinion if it should be in or not. 

Thank you for your future feedback

0xdt85 commented:
Hello,

Was this abandoned ?

Many thanks

skoudoro commented:
Hi @0xdt85,

it is not ready to be merged. See below what is missing:

- Add unit tests (easy and fast)
- Correct PEP8 errors (easy and fast)
- multiple reviews round (medium)
- Theory bias (priority to check)

your review and feedback (theory or code) is welcomed! Please let us know if you have any question concerning theory or code and help us validated this model. We cannot move forward without validation of the community

kikiluvbrains commented:
Hi all, is anyone still working on this? If not, I'm happy to take over, fork the branch, add proper unit tests, and have all errors corrected.

arokem commented:
Hey @kikiluvbrains : glad for you take over. I did some more work in https://github.com/nrdg/fwe/pull/8, but did not get all the way there yet.

kikiluvbrains commented:
Alright, I will follow your methodology  [nrdq/fwe#8](https://github.com/nrdg/fwe/pull/8) and have it integrated natively in DIPY. I see you are  using a a regularized gradient descent free-water elimination model similar to what is mentioned in Pasternak et al., 2009.

arokem commented:
That's right. This is based closely on @mvgolub's implementation.
